### Упражнения по Numpy

In [19]:
import numpy as np

1. Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8 (не включительно)

In [ ]:
n = int(input("Input the size of array:"))
k = int(input("Input max possible integer value in array:"))

mas = np.random.randint(0,k,n)

print("input array: " + str(mas))

for i in range(n):
    if (mas[i] > 3) & (mas[i] < 8):
        mas[i] = - mas[i]


print("output massive: " + str(mas))

2. Заменить (первый) максимальный элемент случайного массива на 0

In [ ]:
n = int(input("Input the size of array:"))
k = int(input("Input max possible integer value in massive:"))

mas = np.random.randint(0,k,n)

print("input array: " + str(mas))

max_elem = mas[0]
max_elem_ind = 0

for i in range(n):
    if mas[i] > max_elem:
        max_elem = mas[i]
        max_elem_ind = i

mas[max_elem_ind] = 0

print("output array: " + str(mas))

3. Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [ ]:
def deck_mul(arrays):
    arrays = tuple(arrays)
    shape = map(len, arrays)

    n_pairs = np.indices(shape, dtype=int)
    n_pairs = n_pairs.reshape(len(arrays), -1).T

    for n, arr in enumerate(arrays):
        n_pairs[:, n] = arrays[n][n_pairs[:, n]]

    return n_pairs

lines = int(input("Input amount of lines in array:"))
columns = int(input("Input amount of columns in array:"))

mas = np.random.randint(10,size = (lines,columns))

print(mas)

print(deck_mul(mas))

4. Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [ ]:
mas_A = np.random.randint(6,size = (8,3))
mas_B = np.random.randint(6,size = (2,2))

print(mas_A)
print(mas_B)

mas_count = np.array([0,0,0,0,0,0,0,0])


for i in range(mas_A.shape[0]):
    k_1: int = 0
    k_2: int = 0
    for j in range(mas_A.shape[1]):
        for p in range(2):
            if(mas_A[i][j]==mas_B[0][p]):
                k_1+=1
            if(mas_A[i][j]==mas_B[1][p]):
                k_2+=1
    if(k_1>=1) & (k_2>=1):
        mas_count[i] = i+1

print("Numbers of rows with elements of ever B_mas row:")
for i in range(mas_A.shape[0]):
    if(mas_count[i]!=0):
        print(mas_count[i])

5. Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [ ]:
matrix = list(np.random.randint(3,size = (10,3)))
print(matrix)

i=9

while i!=-1:
    if(len(set(matrix[i]))<=1):
        matrix.pop(i)
    i-=1

print(matrix)

6.Дан двумерный массив. Удалить те строки, которые повторяются

In [ ]:
matrix = np.random.randint(2,size = (10,3))
print(matrix)

matrix = np.array(matrix)

matrix, idx = np.unique(matrix, axis=0, return_index=True)
res = matrix[idx.argsort()]

print(res)

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).

**Замечание 1.** Можно считать, что все указанные объекты непустые (к примеру, в задаче 1 на диагонали матрицы есть ненулевые элементы).

**Замечание 2.** Для большинства задач решение занимает не больше 1-2 строк.

**Задача 1:** Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.
Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [ ]:
rows:int = int(input("Input number of rows in matrix:"))
columns:int = int(input("Input number of columns in matrix:"))

matrix = np.random.randint(3,size = (rows,columns))
print(matrix)

#realisation without NumPy
matrix = list(matrix)
mul:int = 1

for i in range(len(matrix)):
    if(matrix[i][i]!= 0):
        mul*=matrix[i][i]

print(mul)

#realisation with NumPy

matrix = np.array(matrix)
diagonal_ = np.array(matrix.diagonal())
print(np.prod(diagonal_[diagonal_!=0]))

**Задача 2:** Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.
Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [ ]:
pip install multiset

In [ ]:
from multiset import Multiset

x = list([1,1,2,5,7,8])
y = list([2,5,8,7,1])

#realisation without NumPy

set_x = Multiset(x)
set_y = Multiset(y)

print(set_x==set_y)

#realisation with NumPy

x = np.array(x)
y = np.array(y)

np.sort(x)
np.sort(y)

print(x == y)

**Задача 3:** Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [ ]:
x = list([6, 2, 0, 3, 0, 0, 5, 7, 0])

#realisation without NumPy

k = list([])
for i in range(len(x)-1):
    if (x[i] == 0):
        k.append(x[i+1])


print(max(k))

#realisation with NumPy

x = np.array(x)
print(np.max(x[1:][(x==0)[:-1]]))

**Задача 4:** Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.
Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [ ]:
x = list([0,0,4,4,4,0,5,5,7,7])

#realisation without NumPy

rle_1 = list([])
rle_2 = list([])
k:int = 1

for i in range(0,len(x)-1):
    if(x[i] == x[i+1]):
        k+=1
    else:
        rle_1.append(x[i])
        rle_2.append(k)
        k=1

rle_1.append(x[len(x)-1])
rle_2.append(k)

rle = [rle_1,rle_2]

print(rle)

#realisation with NumPy

x = np.array(x)


def run_length_encoding_numpy(seq_array):
    rle_1 = []
    rle_2 = []
    change = (seq_array[1:] != seq_array[:-1])
    k = np.append(np.where(change), len(seq_array) - 1)
    counter = np.diff(np.append(-1, k))

    for i in range(0, len(counter)):
        rle_1.append(seq_array[k][i])
        rle_2.append(counter[i])

    rle = [rle_1, rle_2]

    return rle

print(run_length_encoding_numpy(x))

**Задача 5:** Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [ ]:
from math import sqrt, dist
from scipy.spatial import distance
import time

In [ ]:
#realisation without NumPy

def python_euclid(X, Y):
    euclid = [[0 for j in range(len(Y))] for i in range(len(X))]

    for i in range(0, len(X)):
        for j in range(0, len(Y)):
            euclid[i][j] = math.sqrt((math.dist(X[i], Y[j])) ** 2)

    return euclid

#realisation with NumPy


def numpy_euclid(X,Y):

    euclid = np.empty(shape=(len(X), len(Y)))

    for i in range(0, len(X)):
        for j in range(0, len(Y)):
            euclid[i][j] = np.linalg.norm(X[i] - Y[j])

    return euclid


#realisation with scipy

def scipy_euclid(X, Y):
    return distance.cdist(X, Y, 'euclidean')

#time tests

X = np.random.randint(0,1000,size=(100,200))
Y = np.random.randint(0,1000,size=(200,200))

start_time_1 = time.time()
python_euclid(X,Y)
print("--- %s seconds ---" % (time.time() - start_time_1))

start_time_2 = time.time()
numpy_euclid(X,Y)
print("--- %s seconds ---" % (time.time() - start_time_2))

start_time_3 = time.time()
scipy_euclid(X,Y)
print("--- %s seconds ---" % (time.time() - start_time_3))

***Задача 6: CrunchieMunchies ***


Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием CrunchieMunchies.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши СrunchieMunchies - самый здоровый выбор для потребителей.

1.Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [ ]:
calorie_stats = np.loadtxt("./sample_data/cereal.csv", delimiter=",")
calorie_stats

В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?
Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [23]:
average_calories = np.average(calorie_stats - 60)
print(average_calories)

46.883116883116884


Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.
Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [25]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [26]:
median_calories = np.median(calorie_stats_sorted)
print(median_calories)

110.0


В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [27]:
percentile = np.empty(shape=101)
for i in range(101):
    percentile[i] = np.percentile(calorie_stats, i)
    if percentile[i]>60:
        nth_percentile = i
        break

print(nth_percentile)

4


Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [28]:
more_calories = len(calorie_stats[calorie_stats>60])/len(calorie_stats)
print(more_calories)

0.961038961038961


Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [29]:
calorie_std = np.std(calorie_stats)
print(calorie_std)

19.35718533390827


Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

CM - CrunchieMunchies

В среднем количество каллорий в CM меньше, чем в остальных злаках на 46 кКал, но так как показатель среднего арифметического подвержен сильному влиянию «больших отклонений», это не может гарантировать, что CM одни из самых лучших в нашей выборке (а что, если в выборке есть суперкалорийные хлопья с превышением суточной нормы кКал в 100 раз, которые рекомендуют 0/10 диетологов, но чёрный пиар тоже пиар, и они продаются?)

Медиана показывает превосходство CM над другими чуть лучше, чем среднее арифметическое, ибо мы можем утверждать, что хотя бы половина злаков содержит в себе на 50 калорий больше, чем CM (однако что, если производители разделились на 2 лагеря? Одни решили сделать вкусные, но калорийные хлопья с содержанием более 100 кКал для настоящих гурманов и сладкоежек, а другие выбрали своей аудиторией фитоняшек и стали выпускать хлопья по 10- кКал, и получилось так, что 39 отнеслись к первому лагерю, а 38 ко второму, тогда наши злаки однозначно входят только в ТОП-39, но не более того).

Перейдём к процентилю. Он показал, что CM однозначно входят в ТОП-4 лучших хлопьев из выборки. Однако понятие процентиля знакомо далеко не всем потребителям (даже я впервые услышала сегодня), поэтому лучше объяснить всё на понятном для большинства языке

А вот процент хлопьев в выборке, в которых содержится более 60 калорий на порцию, довольно высок - 96%, это показывает, что CM находятся в очень выгодной позиции по калорийности. Да, не 100%, но это можно сослать на то, что наши злаки вкуснее этих 50 кКалорийных, поэтому пришлось проиграть по калориям, ведь главное чтобы было не только полезно, но и вкусно

Стандартное отклонение показывает нам, что впринципе, разброс у значений калорийности не является слишком большим, но также и не является слишком маленьким, так что нет значительно выбивающихся за рамки разумного по калорийности злаков, что показывает состоятельность оценки средним арифметическим, в совокупности эти 2 значения тоже можно исползовать как убедительный аргумент.

В целом это всё, что можно было использовать. 

P.S. Простите, некуда было деть свой писательский талант и знание ТВиМСов

